File: packages/wasmvm/wasmlib/ts/wasmlib/state.ts
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

import * as wasmtypes from "./wasmtypes"
import {IKvStore} from "./wasmtypes"
import {stateDelete, stateExists, stateGet, stateSet} from "./host";

export class ScImmutableState {
    exists(key: u8[]): bool {
        return stateExists(key);
    }

    get(key: u8[]): u8[] {
        const val = stateGet(key);
        return val === null ? [] : val;
    }
}

export class ScState implements IKvStore {
    public static proxy(): wasmtypes.Proxy {
        return new wasmtypes.Proxy(new ScState());
    }

    delete(key: u8[]): void {
        stateDelete(key);
    }

    exists(key: u8[]): bool {
        return stateExists(key);
    }

    get(key: u8[]): u8[] {
        const val = stateGet(key);
        return val === null ? [] : val;
    }

    public immutable(): ScImmutableState {
        return this;
    }

    set(key: u8[], value: u8[]): void {
        stateSet(key, value);
    }
}

File: packages/wasmvm/wasmlib/go/wasmlib/state.go
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

package wasmlib

import (
	"github.com/iotaledger/wasp/packages/wasmvm/wasmlib/go/wasmlib/wasmtypes"
)

type ScImmutableState struct{}

func (d ScImmutableState) Exists(key []byte) bool {
	return StateExists(key)
}

func (d ScImmutableState) Get(key []byte) []byte {
	return StateGet(key)
}

type ScState struct {
	ScImmutableState
}

func (d ScState) Bytes() []byte {
	panic("ScState.Bytes: cannot encode state")
}

var _ wasmtypes.IKvStore = ScState{}

func (d ScState) Delete(key []byte) {
	StateDelete(key)
}

func (d ScState) Immutable() ScImmutableState {
	return ScImmutableState{}
}

func (d ScState) Set(key, value []byte) {
	StateSet(key, value)
}

func NewStateProxy() wasmtypes.Proxy {
	return wasmtypes.NewProxy(new(ScState))
}
